### ISSUE TITLE
Notes and questions on logs from reading

### ISSUE THREAD

**Issue by Cleop** [dwyl/phoenix-ecto-append-only-log-example#5]
As someone who is new to logs I could follow the example. However whilst discussing the subject area with others it brought up key terms and subject areas that were new to me. I think it would be useful to include some of this context in the readme for those who may stumble upon this repo without knowing what it is first.

#### What is a log?
**AKA**  write-ahead log, commit log, transaction log. In this repo it **will _not_ refer to** 'application logging', the kind of logging you might see for error messages.

A log is one of the most simple possible storage abstractions. An append-only, totally-ordered sequence of records ordered by time. They are visualised horizontally from left to right.

They're not all that different from a file or a table. If we consider a file as an array of bytes and a table as an array of records. Then a log can be thought of as a kind of table where records are sorted by time. 

Logs are **event driven**. They record what happened and when continuously. As the records are stored in the order that the changes occurred this means that at any point you can revert back to a given point in time by finding it in your records. They can do this in near real-time, making them ideal for analytics. They are also helpful in the event of crashes or errors as their record of the state of the data at all times means data can easily be restored. By keeping an immutable log of the history of your data it means your data is kept clean and is never lost or changed. The log is added to by publishers of data and used / acted upon by subscribers but the records themselves cannot be mutated.

#### Keywords

**Time series database:** a database system optimised for handling time series data (arrays of numbers indexed by time). They handle queries for historical data/ time zones better than relational dbs.

**Data integration:** making all the data an organisation has available in all its services and systems.

**Log compaction:** methods to tidy up a log by deleting no longer needed data.

 #### Questions
- Are we performing physical (the data itself) or logical (the command or calculation which results in the data) logging?
- Are all logs time series databases?
- Does a log contain all of the fields of data that would be captured in a relational DB schema? Does this lead to there being a lot of empty rows because some columns aren't applicable to everything?
- Are all logs append only?
- What is a record in the context of a log? Is it the equivalent of a row in a table? Does it have column titles so all records have the same field titles?
- Do all logs run horizontally?
- Is a timestamp value mandatory in a record in a log? Do they act as the unique identifier for records or does a numeric id?
- When are horizontal scaling partitions useful? I didn't understand in what context they'd be used from the article... Would you have a separate log per user and the partition is made for each user ID?
- The article referred a lot to 'distributed data systems' - if a log is a move away from them, what kind of system would you call a log system? Does it have a name? Is it an integrated system?

**These notes and questions came from reading:**
https://engineering.linkedin.com/distributed-systems/log-what-every-software-engineer-should-know-about-real-time-datas-unifying

**Comment by nelsonic** [comment]
Hi @Cleop, _thank you_ for opening this issue and summarising your knowledge quest! 🎉 

_Stoked_ that you found the Example created by @Danwhy clear and followed it on your `localhost`. 🥇 

# _Answers_ to your Questions (_above_)

+ [ ] "**Physical** vs **Logical**?" >> ***Logical***. 
  We are storing **all data** "CRUD" operations as ***new*** rows in a Postgres Database table 
and then (_much later in the lifecycle of the product_) computing a "View" https://en.wikipedia.org/wiki/View_(SQL) on that data to optimise SELECT queries. 
  > This is _way_ too much detail (_complexity_) for a _beginner_ to worry (_or even know/think_) about, 
but the curious can read:
  > + https://en.wikipedia.org/wiki/Physical_data_model
  > + https://en.wikipedia.org/wiki/Logical_data_model

+ [ ] **All logs** are **time-series _by definition_**. If a record is stored with a timestamp, it's a time series. 
Perhaps a _useful_ clarification/distinction from the Wikipedia article https://en.wikipedia.org/wiki/Time_series_database (_which is kinda useless/confusing_) ... 😞 
_Any_ data can be stored as time series and _most_ ("_big_") data is! 
  > A "beginner-friendly" (_+funny!_) intro to "**_Big_ Data**": https://www.bbc.co.uk/programmes/b0b9wbf8
In the Apps we're building, we're using a [**UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier) as the Primary Key (PK) to avoid write conflicts.
Even with Erlang's Timestamp precision being _microseconds_ anything more than 50 DB writes per second would result in an _unacceptably_ high chance of PK collision. 💔 
**Timestamps** in **`Elixir`**: 
  https://michal.muskala.eu/2017/02/02/unix-timestamps-in-elixir-1-4.html

+ [ ] "**Are all logs append only**?" >> Yes, all logs worth knowing/using are _append-only_.
  If logs are anything _other_ than _append-only_ they are a source of chaos/confusion. 😿 
  > Good "further reading" on this topic:
  > + https://www.confluent.io/blog/using-logs-to-build-a-solid-data-infrastructure-or-why-dual-writes-are-a-bad-idea/
  > + https://security.stackexchange.com/questions/15189/security-issues-with-an-append-only-log-in-unix
  > + https://www.coursera.org/lecture/cryptocurrency/bitcoin-as-an-append-only-log-be6cd

+ [ ] "**What is a record** in the context of a log? Is it the equivalent of a row in a table?" >> **Yes**.
All records are rows. All rows have column headings. Most fields after the _essential_ are _optional_. 
For example in an **Address Book** there's no point having an **_empty_ row** 
(i.e. if all are fields optional, blank records would be "OK", this is obviously undesirable), 
***so*** the _minimum_ data acceptable for a new Address Book entry is the person's `Name`;
_everything_ is optional and can be _added_ on subsequent writes/updates.

+ [ ] "**Do all logs run _horizontally_**?" >> Don't confuse _yourself_ with the _direction_ of "travel". 
All Time is **linear** and **_uni_-directional**. If in _any_ doubt, logs run **_Vertically_ Top-to-bottom** like reading a "page".
  > Think about how you like to scroll a screen/phone using your keyboard, mouse or finger:
  How often do you _scroll **horizontally**_? (_almost **never**. it's a UX anti-pattern_) 
  By contrast scrolling vertically is a _natural_ UX. 
  There might be the _occasional_ timeline in a history book that is horizontal for aesthetic reasons. But it almost always means they have "truncated" the data points to fit on the page.
  Almost no computer systems are horizontal-scrolling. 
  **Note**: horizontal scrolling is **not** the same as "swiping" on a mobile; the go-to UX for all hipster apps: https://uxplanet.org/horizontal-scrolling-in-mobile-643c81901af3

+ [ ] "Is a **timestamp** value **mandatory** in a record in a log?" >> Yes, by definition. 
And thankfully, Phoenix/Ecto _already_ stores timestamps: **`inserted_at`**. 🎉 
  + [ ] Do they act as the unique identifier for records or does a numeric id? >> No.
  > Timestamps should _never_ be used as a unique identifier in any _serious_ system. 
  The Primary Key (PK) should be _Universally_ Unique (_hence using using a [**UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier)_) i.e. virtually zero chance of conflict.
  Using a Timestamp would almost _always_ **guarantee conflict** where writes-per second 
  are greater than 1k. In a _modest_ "Chat" application with 100k concurrent users, 
  1k writes/sec are "normal". (_each person sends one message once every couple of minutes..._) 
  > See: http://www.internetlivestats.com
![time](https://thumbnails-visually.netdna-ssl.com/the-internet-in-real-time_53ac20c056172.gif)

+ [ ] "**_horizontal scaling_** partitions useful?" >> 😄 (_What will you do when you win the lottery?_)
  (_it's good that you are keen! but you are over-thinking this ... com back to this question in 2020!_) 
  Thinking about Horizontal Scaling is _way_ beyond the "scope" of this example/tutorial. 
  To be _clear_: we are using an Append-only Log for reliability, accountability and record rollback-ability, not "scalability" (i.e. "premature optimisation"). Scalability is a _distant future_ benefit, not a focus.
   > If you (_or anyone else_) are _personally_ interested in Scalability you should read: https://stackoverflow.com/questions/11707879/difference-between-scaling-horizontally-and-vertically-for-databases
  Using an Append-only Log with UUIDs as PKs is _all_ the "ground work" we need 
  to ensure that _anything_ we build is _prepared_ to scale **_both_ Vertically _and_ Horizontally**. ✅ 🚀 
  When _any_ of our apps reaches **10k writes/sec** we will be **_insanely_** "**successful**". 🦄 🎉 
  An AWS RDS [`db.m4.16xlarge` instance](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Concepts.DBInstanceClass.html) has **256GB** of RAM and can handle 10GB of "throughput". It's been _benchmarked_ at ***200k writes/second*** ... if we _ever_ need to use **one** of these instances, we'll all be sipping coconut water on the shore of the @dwyl [island](https://github.com/nelsonic/nelsonic.github.io/issues/341)! 🌴☀️🍸 🏄‍♀️ 🐢 🐟 ⛵️  

+ [ ] '**distributed data systems**' - "_if a log is a **move away** from them_" ...? >> Huh...? 😕 
![log-solves-distributed-data](https://user-images.githubusercontent.com/194400/46251765-8f94f100-c454-11e8-8f2f-d56bba8af0b3.png)
  Logs are the _basis_ for all distributed/fault-tolerant systems. That's all we need to know/say.
  > Please avoid adding more terms/complexity to the readme than are _strictly_ necessary; 
anything more than "append-only log" will _automatically_ "confuse" people who new to this. 
  If you are _curious_ about _understanding_ this in _way_ more detail, read: https://raft.github.io
  and watch [@substack](https://github.com/substack)'s talk on "What you can build with a log":
[![image](https://user-images.githubusercontent.com/194400/46251788-42fde580-c455-11e8-98a0-06219f42c44c.png)](https://youtu.be/RPFjN1N148U)
https://youtu.be/RPFjN1N148U
For the purposes of _all_ developers using Phoenix, just _trust_ PostgreSQL to handle your data.
You're in "good company" https://github.com/dwyl/learn-postgresql/issues/31 
And if our app is _lucky_ enough to be "successful", look into [CitusDB](https://github.com/dwyl/how-to-choose-a-database/issues/4)

There is no need to _confuse_ beginners with the term "_write-ahead log_" https://en.wikipedia.org/wiki/Write-ahead_logging (["TMI"](https://en.wiktionary.org/wiki/too_much_information)); it's just an append-only log. 
Hopefully people will understand the words: ["append"](https://en.wikipedia.org/wiki/Append), ["only"](https://en.wiktionary.org/wiki/only#Conjunction) and ["log"](https://en.wikipedia.org/wiki/Log_file) ... 💭
(_if not, they probably aren't "ready" for this example/tutorial ..._)

In general/practice, **log compaction** is **_never_ needed** for _most_ companies/products these days.
Data storage costs are so _cheap_ (_see detail below_) that every App developer can _easily_ afford to store _all_ data generated by their App - provided the app does not store arbitrary data for no reason.
In most companies, data is the "life blood" of the strategic decision making, the entire field of "BI" https://en.wikipedia.org/wiki/Business_intelligence relies on having as _much_ data as possible. 

> ### _Aside_: _Never_ Over-write or Delete Data it is your _Most Valuable Asset_!
> I feel "qualified" to assert that using Append-only Logs for _everything_ in a company
> are the _single_ best technology decision that can be made. Logs are the _foundation_ for the building.
> I helped setup "BI" at Groupon, no data was _ever_ deleted or compacted, only periodically "warehoused" to save on storage cost, but still accessible via Hadoop. 
Per employee, we were the _single_ biggest source of additional revenue, cost savings and value creation in the company. A "department" of 8 people generated $20M of annual value for the company by spotting trends in markets, products & buying behaviour. 📈 
Hedge funds (_like [Bridgewater](https://en.wikipedia.org/wiki/Bridgewater_Associates) or [Citadel](https://en.wikipedia.org/wiki/Citadel_LLC)_) use data analysis to **_multiply_ cash**. 
It's a _fascinating_ world/activity that generates no (_NET_) value to society 
but makes a _handful_ of people _phenomenally_ wealthy! 
https://en.wikipedia.org/wiki/The_rich_get_richer_and_the_poor_get_poorer

> If you're ever interested in learning how _some_ Mathematicians use their skills to **_print_ money**, 
> read: https://en.wikipedia.org/wiki/The_Quants (_I think we still have a copy in the @dwyl library..._) 
> Or if you're short on time or prefer to watch the dramatised version, [The "Big Short"](https://en.wikipedia.org/wiki/The_Big_Short ) is good. 📺 
If you haven't seen it, add it to your list: https://www.imdb.com/title/tt1596363
***Or*** [Spoiler Alert] _just_ watch the _ending_: https://youtu.be/Bu2wNKlVRzE
The _most_ insightful part of the film is in the **_end_ credits**. [Michael Burry](https://en.wikipedia.org/wiki/Michael_Burry) is investing in [***Water***](https://github.com/dwyl/phase-two/issues/21):
[![image](https://user-images.githubusercontent.com/194400/46250986-feffe600-c43e-11e8-875d-956d7582fac7.png)](https://github.com/dwyl/phase-two/issues/21)
If you or anyone else has "spare capital", to do the same; invest in **water** (_recovery/treatment_).
(_**not** to profit from gouging the poor! but **water treatment startups/ideas/tech** are ["**win-win**"](https://en.wikipedia.org/wiki/Win-win_game)_)


# tl;dr

## _Some_ Data gets Updated, _Most_ Never Does

Many developers  have an "_observation bias_" toward data changing (_mutability_) because we often change the data we _interact_ with and we therefore _think_ that mutability is the "norm" but it _really_ is not.
Once people realise the fact that **_most_ data** gets created/written _once_ and **_never_ gets updated**. 
Append-only logs are (_unsurprisingly_) common in virtually every area of computing.

Even when data _appears_ to be mutable (_from the User's perspective_) it is often stored in an immutable log so the underlying store is immutable but the UI/UX _appears_ that data is being mutated because the UI only displays the _latest_ version of the data.

In many popular web frameworks, preserving "history" of a record (_or piece of content_) 
is often done with a "history table" which stores copies of the "previous" versions of data whenever a record is updated. 

## _Examples_ in Every Industry

Examples of use-cases where Append-only Logs are useful were included 
in the original "Why? What? How?" issue: https://github.com/dwyl/phoenix-ecto-append-only-log-example/issues/1

To those examples we can _add_:

+ **Banking/Financial transactions** are _all_ append-only (_write once_) ledgers. If they were not, the all accounting would be chaos and the world economy would collapse. When the "available balance" of an account is required, it is _calculated_ from the list/log of transactions. 
(_a summary of the data in an account may be **cached** in a "view" but it is **never mutated**_)
  > You may recall at the End of your F&C course you were applying to work for a company and they asked you to do a coding challenge/exercise that involved computing the values of a bank account.
We paired on solving that challenge using a Log (debit and credit ledger). Data was never mutated. The **account balance** was _always **calculated**_ "just in time" by "traversing" the Log.
+ ***Healthcare***: a patient's medical data gets captured/recorded once as a "snapshot" in time. The doctor or EKG machine does not go back and "update" the value of the patients heart rate or HDL level. A _new_ value is sampled at each time interval.
+ **Analytics** is _all_ append-only logs which are (time) series of events _streamed_ from the device to server, saved in a time-series data store, and streamed (_or "replayed"_) to visualisation dashboard.
  + Events in Analytics systems are often _aggregated_ (_using "views"_) into charts/graphs. The "views" of the data are "temporary tables" which store the _aggregated_ or _computed_ data but do not touch the underlying log/stream.

Fact is: that append-only logs while not often mentioned by name, are in fact the _norm_.
Think of an application that stores data of any kind, and you can see how it can either _benefit_ from
(or is _totally dependent_ on) data immutability.

> ***Everything*** is ***always*** an **event**. 
Everything in the _universe_ happens _once_. Time is _linear_ and _uni-directional_. 
Even in the ["multiverse hypothesis"](https://en.wikipedia.org/wiki/Multiverse), time is still uni-directional, we cannot (_no theory in quantum mechanics allows us to_) **go _back_ in time**;
**_Mutating_ data** is "**time travelling**"; it's updating data that was created in the past and _destroying_ the history/timeline. In a distributed system mutating data means unpredictability/unreliability.


### Address Book Example

In the example given in the `README.md` of an Address Book, 
we illustrate how a person can "_update_" their address when they move home, 
_however_ (_in the example_) their _previous_ address does not get _over-written_,
we simply insert their _new_ address into the database 
and the _new_ address is treated as the "current" version.
That way the "Address Book" has a complete record of the history. 
This is a _good_ use of an append-only log 
because address data will almost _always_ change over the lifetime of a person,
 but knowing previous addresses is useful (_and in some cases **essential**_).

If **Thor** decides to leave his parents' house ([planet](https://en.wikipedia.org/wiki/Asgard)) 
His "_old_" address for reference:
```
Thor
The Hall, Valhalla, Asgard
AS1 3DG
0800123123
```
![thor-old-address](https://user-images.githubusercontent.com/194400/46251421-08db1680-c44a-11e8-95cc-ce8815ef7d18.png)
and crashes on his buddy [Stephen's](https://en.wikipedia.org/wiki/Doctor_Strange) couch till he gets his own place in NY.
His _new (_temporary_) address will be:

```
Thor Odinson c/o Dr. Strange
177A Bleecker Street,
New York, NY 10012, USA

```
![thor-new-address](https://user-images.githubusercontent.com/194400/46251403-bc8fd680-c449-11e8-9b1f-73d137dd68c0.png)

Thanks @mathiasbynens for this handy character/byte counter: https://mothereff.in/byte-counter 🥇 

> The _average_ person in the United States is expected to move 11.4 times in their lifetime:
https://fivethirtyeight.com/features/how-many-times-the-average-person-moves


### Historical Context of _Mutable_ Data: It _Was Too Expensive_ to Store _Everything_

To understand ***why*** _many_ Web Application frameworks (_still_) _over-write_ data when an **`update`** is made,
we need to look back to when data _storage_ was _expensive_. At the dawn of digital computing 
storing data was _prohibitively_ expensive. In 1956 a storing a Megabyte of data cost **$9,200**.
i.e. **$9.2 Million _per_ Gigabyte**. IBM charged **$850 a month** (_rent_) for **3.75** Megabytes of storage. 
![image](https://user-images.githubusercontent.com/194400/46250682-c0b2f880-c437-11e8-9937-fb151bd44a8e.png)
With a Megabyte of storage, and assuming that the average (home) address requires **50 bytes** we could store **20K addresses**. This is _barely_ enough to store the addresses for for a _town_. 
When the cost of storage is _this_ expensive storing all previous versions is _unfeasible_.
This is _also_ why _early_ mobile phones had a limited contact list (_100 contacts in most cases_)

_Today_ we can store an _incredible_ amount of data on a **MicroSD** card the size of a postage stamp: 
![image](https://user-images.githubusercontent.com/194400/46241883-4beca980-c3b8-11e8-8f95-9f0a35ddddb2.png)
Data storage (_cost_) is no longer a limiting factor, so we can _afford_ to design _all_ applications to be immutable.

> For complete history, see: http://www.computerhistory.org/timeline/memory-storage
> For relative prices of data storage see: https://jcmit.net/diskprice.htm
> For a good intro to Bits and Bytes, see: https://web.stanford.edu/class/cs101/bits-bytes.html
